### PR TITLE
Add a flag to disable parallel configuration in IP mode in favor of COD

### LINKED ISSUE #30401: Provide a flag to disable parallel configuration for IP mode in favor of configure-on-demand
Enabling configuration parallelism in IP mode without incrementality is intentional tradeoff that been taken. It will allow to observe performance improvements been brought by parallelism during IDE project sync and tasks running from CLI.

However, for the builds of the extreme size (in terms of projects count), benefits from configure-on-demand could be more significant than benefits from parallel configuration.

### PR DESCRIPTION
**Pull request by 6hundreds**
Fixes #30401

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bot-gradle test this without pts on linux

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt [hunk 1]
```diff
@@ -141,7 +145,7 @@ class DefaultBuildTreeModelControllerServices : BuildTreeModelControllerServices
             )
         } else {
             val configurationCache = isolatedProjects || startParameter.configurationCache.get()
-            val configureOnDemand = !isolatedProjects && startParameter.isConfigureOnDemand
+            val configureOnDemand = (!isolatedProjects || isolatedProjects && options.getOption(isolatedProjectsTasksConfigureOnDemand).get()) && startParameter.isConfigureOnDemand
```
Comment (6hundreds) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt:148]: Or do we say that `isolatedProjectsTasksConfigureOnDemand ` is enough, no `startParameter.isConfigureOnDemand` respect is needed?

Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt:148]:
```suggestion
            val configureOnDemand = (!isolatedProjects || options.getOption(isolatedProjectsTasksConfigureOnDemand).get()) && startParameter.isConfigureOnDemand
```

**Review by bamboo — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt (hunk 1 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt]: Enabling IP should eventually imply incrementality, which implies `val configureOnDemand` should be true. In this PR we are saying that for the IP to imply incrementality (when running tasks) an extra internal flag should be enabled.

For all the brevity of the complex logical expressions, they often become cryptic, hiding the original intent.
My take on the condition would be:
```
val configureOnDemand = if (isolatedProjects) new_internal_flag else startParameter.configureOnDemand
```

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy [hunk 2]
```diff
@@ -66,6 +65,18 @@ class IsolatedProjectsParallelConfigurationIntegrationTest extends AbstractIsola
         [":build", "--no-configure-on-demand"] | [":build"]
     }
 
+    def 'parallel configuration can be disabled in favor of configure-on-demand'() {
+        given:
+        server.expect("configure-root")
+        server.expect("configure-a")
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy:71]: 💭  The server here waits for some heuristic time (500ms?) to ensure no further requests are received. Though, I don't think it makes it clear that `:b` was not evaluated at all. I'd make it more explicit.

Ideally, we should be checking whether the project has been evaluated via build operations. But a println in `:b`'s build logic and the absence of it in the output should suffice.

**Review by alllex — CHANGES_REQUESTED** [review]
See my comment for the condition, which suggests altering it to not being dependant on the `--configure-on-demand`, when IP is enabled.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt (hunk 1 above)
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt]: The `if` formulation is clear 👍

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy (hunk 2 above)
Comment (6hundreds) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy:71]: Printing doesn't allow to verify sequentiality of configuration tho

Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy:71]: My suggestion is only to add a validation, not remove the ones already present

**Review by alllex — APPROVED** [review]
LGTM!

**Review by bamboo — APPROVED** [review]

**Pull request merged**